src/node/rpc/jsonhandler.h

```diff
@@ -0,0 +1,301 @@
+   *   }
+   *   else
+   *   {
+   *     ...
+   *   }
+   *   auto result = fn(params);
+   *   if (is_error(result))
+   *   {
+   *     args.rpc_ctx->set_response_status(SOME_ERROR);
+   *     args.rpc_ctx->set_response_header(content_type, Text);
+   *     args.rpc_ctx->set_response_body(error_msg(result));
+   *   }
+   *   if (pack_type == Text)
+   *   {
+   *     args.rpc_ctx->set_response_header(content_type, JSON);
+   *     args.rpc_ctx->set_response_body(pack(result, Text));
+   *   }
+   *   else
+   *   {
+   *     ...
+   *   }
+   * };
+   *
+   * it is possible to write the shorter, clearer, return-based lambda:
+   * auto foo = json_adapter([](Store::Tx& tx, nlohmann::json&& params)
+   * {
+   *    auto result = fn(params);
+   *    if (is_error(result))
+   *    {
+   *      return make_error(SOME_ERROR, error_msg(result));
+   *    }
+   *    else
+   *    {
+   *      return make_success(result);
+   *    }
+   * });
+   */
+
+  namespace details
+  {
+    struct ErrorDetails
+    {
+      http_status status;
+      std::string msg;
+    };
+
+    using JsonAdapterResponse = std::variant<ErrorDetails, nlohmann::json>;
+
+    static constexpr char const* pack_to_content_type(jsonrpc::Pack p)
+    {
+      switch (p)
+      {
+        case jsonrpc::Pack::Text:
+        {
+          return http::headervalues::contenttype::JSON;
+        }
+        case jsonrpc::Pack::MsgPack:
+        {
+          return http::headervalues::contenttype::MSGPACK;
+        }
+        default:
+        {
+          return nullptr;
+        }
+      }
+    }
+
+    static jsonrpc::Pack detect_json_pack(
+      const std::shared_ptr<enclave::RpcContext>& ctx)
+    {
+      std::optional<jsonrpc::Pack> packing = std::nullopt;
+
+      const auto content_type_it =
+        ctx->get_request_header(http::headers::CONTENT_TYPE);
+      if (content_type_it.has_value())
+      {
+        const auto& content_type = content_type_it.value();
+        if (content_type == http::headervalues::contenttype::JSON)
+        {
+          packing = jsonrpc::Pack::Text;
+        }
+        else if (content_type == http::headervalues::contenttype::MSGPACK)
+        {
+          packing = jsonrpc::Pack::MsgPack;
+        }
+        else
+        {
+          throw std::logic_error(fmt::format(
+            "Unsupported content type {}. Only {} and {} are currently "
+            "supported",
+            content_type,
+            http::headervalues::contenttype::JSON,
+            http::headervalues::contenttype::MSGPACK));
+        }
+      }
+      else
+      {
+        packing = jsonrpc::detect_pack(ctx->get_request_body());
+      }
+
+      return packing.value_or(jsonrpc::Pack::Text);
+    }
+
+    static nlohmann::json get_params_from_body(
+      const std::shared_ptr<enclave::RpcContext>& ctx, jsonrpc::Pack pack)
+    {
+      return jsonrpc::unpack(ctx->get_request_body(), pack);
+    }
+
+    static nlohmann::json get_params_from_query(
```

Comment:
This is an early sketch of what this might look like, and needs some more thought. Basically for `GET` requests, we should be able to write `/users/LOG_get?id=42` rather than `{"id": 42}` in the body. This does that parsing, but we need to decide how complicated we allow these queries to be, how they're escaped, and how this hooks into the OpenAPI spec (each endpoint should declare whether it expects the params in the query or in the body).

Reply:
I think we should maybe look at off-the-shelf options, this one has tests and is allocation-free for example: https://github.com/jacketizer/libyuarel/blob/master/yuarel.h

Reply:
I agree we should look, but I don't think there's anything that gets us very far. `yuarel` for instance is merely splitting at `&`s - we can do that trivially, and what we're really interested in is handling escaped-values and type-converting the returned strings.

Reply:
You’re right, but perhaps we can at least get some tests that way.